elk/evaluation/evaluate.py

```diff
@@ -0,0 +1,112 @@
+from ..files import elk_reporter_dir, memorably_named_dir
+from ..utils import assert_type, held_out_split, int16_to_float32, select_usable_devices
+
+
+@dataclass
+class EvaluateConfig(Serializable):
+    target: ExtractionConfig
+    source: str = field(positional=True)
+    normalization: Literal["legacy", "none", "elementwise", "meanonly"] = "meanonly"
+    max_gpus: int = -1
+
+
+def evaluate_reporter(
+    cfg: EvaluateConfig,
+    dataset: DatasetDict,
+    layer: int,
+    devices: list[str],
+    world_size: int = 1,
+):
+    """Evaluate a single reporter on a single layer."""
+    rank = os.getpid() % world_size
+    device = devices[rank]
+
+    # Note: currently we're just upcasting to float32 so we don't have to deal with
+    # grad scaling (which isn't supported for LBFGS), while the hidden states are
+    # saved in float16 to save disk space. In the future we could try to use mixed
+    # precision training in at least some cases.
+    with dataset.formatted_as("torch", device=device, dtype=torch.int16):
+        train, test = dataset["train"], held_out_split(dataset)
+        test_labels = cast(Tensor, test["label"])
+
+        _, test_h = normalize(
+            int16_to_float32(assert_type(Tensor, train[f"hidden_{layer}"])),
+            int16_to_float32(assert_type(Tensor, test[f"hidden_{layer}"])),
+            method=cfg.normalization,
+        )
+
+    reporter_path = elk_reporter_dir() / cfg.source / "reporters" / f"layer_{layer}.pt"
+    reporter = torch.load(reporter_path, map_location=device)
+    reporter.eval()
+
+    test_x0, test_x1 = test_h.unbind(dim=-2)
+
+    test_result = reporter.score(
+        (test_x0, test_x1),
+        test_labels,
+    )
+
+    stats = [layer, *test_result]
+    return stats
+
+
+def evaluate_reporters(cfg: EvaluateConfig, out_dir: Optional[Path] = None):
+    ds = extract(cfg.target, max_gpus=cfg.max_gpus)
+
+    layers = [
+        int(feat[len("hidden_") :])
+        for feat in ds["train"].features
+        if feat.startswith("hidden_")
+    ]
+
+    devices = select_usable_devices(cfg.max_gpus)
+    num_devices = len(devices)
+
+    transfer_eval = elk_reporter_dir() / cfg.source / "transfer_eval"
+    transfer_eval.mkdir(parents=True, exist_ok=True)
+
+    if out_dir is None:
+        out_dir = memorably_named_dir(transfer_eval)
+    else:
+        out_dir.mkdir(parents=True, exist_ok=True)
+
+    # Print the output directory in bold with escape codes
+    print(f"Saving results to \033[1m{out_dir}\033[0m")
+
+    with open(out_dir / "cfg.yaml", "w") as f:
+        cfg.dump_yaml(f)
+
+    cols = ["layer", "loss", "acc", "cal_acc", "auroc"]
+    # Evaluate reporters for each layer in parallel
+    with mp.Pool(num_devices) as pool, open(out_dir / "eval.csv", "w") as f:
+        fn = partial(
+            evaluate_reporter, cfg, ds, devices=devices, world_size=num_devices
+        )
+        writer = csv.writer(f)
+        writer.writerow(cols)
```

Comment:
Could you change this so that it mirrors how we're doing the progress bar and CSV writing in the latest commit?
Currently we use `imap_unordered` but then write to the CSV all at once inside a try-finally block to make sure it happens even when we crash